pkg/encode/encoder.go

```diff
@@ -0,0 +1,451 @@
+		return err
+	}
+	if err := e.WriteInt(ref); err != nil {
+		return err
+	}
+	return nil
+}
+
+// 写入字段信息
+func (e *Encoder) writeFields(fields []*Field) error {
+	for _, field := range fields {
+		if err := e.writeField(field); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// 写入单个字段信息
+func (e *Encoder) writeField(field *Field) error {
+	if err := e.WriteString(field.Name); err != nil {
+		return err
+	}
+	if err := e.writeTypeRefOrClass(field.Type); err != nil {
+		return err
+	}
+	if field.Default != nil {
+		if err := e.WriteObject(field.Default); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// 写入类型引用或类名
+func (e *Encoder) writeTypeRefOrClass(clazz *Class) error {
+	if ref, ok := e.findTypeRef(clazz); ok {
+		if err := e.writeTypeRef(ref); err != nil {
+			return err
+		}
+	} else {
+		if err := e.writeClass(clazz); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// 获取结构体字段信息
+func getClassFields(t reflect.Type) ([]*Field, error) {
+	var fields []*Field
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		name := strings.ToLower(field.Name)
+		typ, err := getClass(field.Type)
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, &Field{Name: name, Type: typ})
+	}
+
+	return fields, nil
+}
+
+func (e *Encoder) WriteNull() error {
+	return nil
+}
+
+func (e *Encoder) findObjRef(obj interface{}) (int, bool) {
+	for i, ref := range e.objRefs {
+		if obj == ref {
+			return i, true
+		}
+	}
+	return -1, false
+}
+
+func (e *Encoder) WriteBool(b bool) error {
+	if b {
+		return e.WriteByte(commons.BC_TRUE)
+	} else {
+		return e.WriteByte(commons.BC_FALSE)
+	}
+}
+
+func (e *Encoder) WriteByte(b byte) error {
+	e.buffer.WriteByte(b)
+	return nil
+}
+func (e *Encoder) writeDouble(f float64) error {
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], math.Float64bits(f))
+	if err := e.WriteByte(commons.BC_DOUBLE); err != nil {
+		return err
+	}
+	_, err := e.buffer.Write(buf[:])
+	return err
+}
+
+func (e *Encoder) WriteInt(i int64) error {
+	if i >= commons.INT_DIRECT_MIN && i <= int64(commons.INT_DIRECT_MAX) {
+		// 一字节表示
+		return e.WriteByte(byte(i) + commons.BC_INT_ZERO)
+	} else if i >= commons.INT_BYTE_MIN && i <= commons.INT_BYTE_MAX {
+		// 两字节表示
+		if err := e.WriteByte(byte(i>>8) + commons.BC_INT_ZERO); err != nil {
+			return err
+		}
+		return e.WriteByte(byte(i))
+	} else if i >= commons.INT_SHORT_MIN && i <= commons.INT_SHORT_MAX {
+		// 三字节表示
+		if err := e.WriteByte(byte(i>>16) + commons.BC_INT_SHORT_ZERO); err != nil {
+			return err
+		}
+		if err := e.WriteByte(byte(i >> 8)); err != nil {
+			return err
+		}
+		return e.WriteByte(byte(i))
+	} else {
+		// 八字节表示
+		if err := e.WriteByte(commons.BC_LONG); err != nil {
+			return err
+		}
+		if err := binary.Write(e.buffer, binary.BigEndian, i); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (e *Encoder) WriteString(s string) error {
+	// 将字符串转换为UTF-8编码的字节序列
+
+	b := []byte(s)
```

Comment:
It will make a copy of the string, which may be slow.
You can use `buffer.WriteString(s)` here; or for a faster conversion from string -> []byte, refer to https://github.com/cloudwego/kitex/blob/develop/pkg/utils/byte2str.go